### ISSUE TITLE
Support OPDS

### ISSUE THREAD

**Issue by kelson42** [kiwix/kiwix-js#489]
OPDS is the new way to deliver the ZIM catalog. See https://wiki.kiwix.org/wiki/OPDS
Would that be possible to support OPDS in Kiwix JS?

**Comment by Jaifroid** [comment]
@kelson42 Three questions about this (related to Kiwix JS Windows, which has an implementation of a download library):

1. Are there plans to deprecate direct browsing of `http://download.kiwix.org/zim/` for directories and filenames? I currently scrape those indices for filenames in Kiwix JS Windows. Not the right way, I know... _NB I get the download links from the `.meta4` file, so users of the app have a choice of mirrors._

2. Currently I can get the torrent for a given ZIM by adding `.torrent` instead of `.meta4` on the filename, and this works. Can I still rely on that? 

3. Have all "split" versions of ZIM files been deprecated (*.zimaa *.zimbb etc.)? There are now only quite old files in `http://download.kiwix.org/zim/portable/`. If so, I should remove my "portable" links.

I know the _right way_ is to query the catalog, but it may take some time to develop...

**Comment by kelson42** [comment]
>     1. Are there plans to deprecate direct browsing of `http://download.kiwix.org/zim/` for directories and filenames? I currently scrape those indices for filenames in Kiwix JS Windows. Not the right way, I know... _NB I get the download links from the `.meta4` file, so users of the app have a choice of mirrors._

No

>     2. Currently I can get the torrent for a given ZIM by adding `.torrent` instead of `.meta4` on the filename, and this works. Can I still rely on that?

Yes

>     3. Have all "split" versions of ZIM files been deprecated (*.zimaa *.zimbb etc.)? There are now only quite old files in `http://download.kiwix.org/zim/portable/`. If so, I should remove my "portable" links.

Yes

**Comment by mossroy** [comment]
Listing and downloading ZIM files from the app is currently not implemented in kiwix-js. We simply point to https://download.kiwix.org/zim/.
@Jaifroid has implemented some logic inside kiwix-js-windows (without OPDS)

**Comment by kelson42** [comment]
The OPDS stream has been strongly developed and as soon as libkiwix10 is release (in a few days) shoukd be considered ready to use.

**Comment by Jaifroid** [comment]
@kelson42 Is the pure OPDS reader implemented in JS, do you know?

**Comment by kelson42** [comment]
@Jaifroid yes, you can try it by downloading a nightly build of kiwix-serve. The js code is in the libkiwix for the moment. I'm perfectly aware that such a piece of software/js library would benefit to many of our solutions and therefore if your are happy with it we could/should consider to isolate it to simplify maintenance/integration.

**Comment by kelson42** [comment]
@Jaifroid This issue desserves an update IMHO

**Comment by Jaifroid** [comment]
@kelson42 We currently frame browse.library.kiwix.org, which is a simple solution that appears to work pretty well in the Browser Extension. There is a fallback for people like RMS whose browsers block library.kiwix.org, which is to a mirror that still serves the text-based view we used to have on donwload.kiwix.org. I'm not sure there's a good argument for reproducing the OPDS natively in this context (desktop browsers). In the PWA, I have a custom text-based OPDS browser that also works very well since updating it to use OPDS (fully) under the hood (rather than download.kiwix.org, as I had previously used), but that makes more sense in that context, since the PWA can be installed on mobile (Android/iOS/Ubuntu Touch, plus deprecated Firefox OS/Windows Mobile) as well as desktop, and screen real estate is at a premium. What would you advise? Do you find the current Library implementation usable in the browser extension? It's not much different to giving the user a link to library.kiwix.org, but framing that does keep the user in the app.

**Comment by kelson42** [comment]
> @kelson42 Three questions about this (related to Kiwix JS Windows, which has an implementation of a download library):
> 
> 1. Are there plans to deprecate direct browsing of `http://download.kiwix.org/zim/` for directories and filenames? I currently scrape those indices for filenames in Kiwix JS Windows. Not the right way, I know... _NB I get the download links from the `.meta4` file, so users of the app have a choice of mirrors._

No plan but it can happen any time. This has never been considered an official interface. As such this will happen - if it happens - without any warning or grace period.

> 2. Currently I can get the torrent for a given ZIM by adding `.torrent` instead of `.meta4` on the filename, and this works. Can I still rely on that? 

No, this is why we work on an OPDS feed listing also the BitTorrent link. Will be live soon.

> 3. Have all "split" versions of ZIM files been deprecated (*.zimaa *.zimbb etc.)? There are now only quite old files in `http://download.kiwix.org/zim/portable/`. If so, I should remove my "portable" links.

Still perfectly supported. Not sure if this part of the specification or only libzim feature set though.

> I know the _right way_ is to query the catalog, but it may take some time to develop...

Yes, nothing is free unfortunately ;)

**Comment by kelson42** [comment]
@Jaifroid Not sure it works honestly: 

<img width="875" height="547" alt="Image" src="https://github.com/user-attachments/assets/44630847-9840-42ab-99dd-730d5fd2416c" />

But it is not like you had a lot of choice: OPDS or library.kiwix.org are the solution.

**Comment by Jaifroid** [comment]
Hmm, it works in Firefox on my machine:

<img width="2322" height="1487" alt="Image" src="https://github.com/user-attachments/assets/ccba4316-7df7-4714-8395-57d43a550d42" />

I wonder what could be blocking it on yours.... Can you give me the version of Firefox and whether it's on Linux or another OS? Also, what's the part of the message that is cut off? It looks like Firefox is saying the issue is the CORS header of browse.library.kiwix.org..., but that should also block mine if it were true...

**Comment by Jaifroid** [comment]
@kelson42 The issue appears to be that browser.library.kiwix.org has blocked framing the library except from specified domains, one of which is browser-extension.kiwix.org . This means that to access the library, you currently need to enable Service Worker mode, whereas you are running your copy in Restricted Mode. In Service Worker mode, it should work. I'll create a new issue for this, because the library should not be blocking the iframe for the local extension.

**Comment by Jaifroid** [comment]
@kelson I've posted #1490 . Where should I report this upstream -- on Operations? The fix is server-side, not in this Repo's code, but I created the issue in case users come looking.

**Comment by Jaifroid** [comment]
I opened https://github.com/kiwix/operations/issues/790 .

**Comment by Jaifroid** [comment]
@kelson42  Upstream has been fixed by rgaudin. Please try again, and confirm whether you see the fix in Restricted mode (it's working for me), and whether you think it's a "good enough" solution for the Browser Extension (as opposed to implementing our own OPDS code in this Repo). If so, this issue can be closed.
